accord-core/src/main/java/accord/messages/ReadData.java

```diff
@@ -18,236 +18,242 @@
-            }
+    @Override
+    protected void process()
+    {
+        waitingOn = new BitSet();
+        node.mapReduceConsumeLocal(this, readScope, executeAtEpoch, executeAtEpoch, this);
+    }
 
-            command.removeListener(this);
-            if (!isObsolete)
-                read(command);
-        }
+    @Override
+    public PreLoadContext listenerPreLoadContext(TxnId caller)
+    {
+        Set<TxnId> ids = new HashSet<>();
+        ids.add(txnId);
+        ids.add(caller);
+        return PreLoadContext.contextFor(ids, keys());
+    }
 
-        @Override
-        public boolean isTransient()
+    @Override
+    public boolean isTransient()
+    {
+        return true;
+    }
+
+    @Override
+    public synchronized void onChange(SafeCommandStore safeStore, Command command)
+    {
+        logger.trace("{}: updating as listener in response to change on {} with status {} ({})",
+                this, command.txnId(), command.status(), command);
+        switch (command.status())
         {
-            return true;
+            default: throw new AssertionError();
+            case NotWitnessed:
+            case PreAccepted:
+            case Accepted:
+            case AcceptedInvalidate:
+            case Committed:
+                return;
+
+            case PreApplied:
+            case Applied:
+            case Invalidated:
+                obsolete();
+            case ReadyToExecute:
         }
 
-        private synchronized void readComplete(CommandStore commandStore, Data result)
-        {
-            logger.trace("{}: read completed on {}", txnId, commandStore);
-            data = data == null ? result : data.merge(result);
+        command.removeListener(this);
+        if (!isObsolete)
+            read(safeStore, command);
+    }
 
-            waitingOn.remove(commandStore);
-            if (waitingOn.isEmpty())
-                node.reply(replyToNode, replyContext, new ReadOk(data));
+    @Override
+    public synchronized ReadNack apply(SafeCommandStore safeStore)
+    {
+        Command command = safeStore.command(txnId);
+        Status status = command.status();
+        logger.trace("{}: setting up read with status {} on {}", txnId, status, safeStore);
+        switch (command.status()) {
+            default:
+                throw new AssertionError();
+            case Committed:
+                if (!command.partialTxn().keys().any((i, k) -> safeStore.commandStore().hashIntersects(k)))
+                    throw new IllegalStateException();
+            case NotWitnessed:
+            case PreAccepted:
+            case Accepted:
+            case AcceptedInvalidate:
+                waitingOn.set(safeStore.commandStore().id());
+                ++waitingOnCount;
+                command.addListener(this);
+
+                if (status == Committed)
+                    return null;
+
+                safeStore.progressLog().waiting(txnId, ReplicationPhase.Commit, readScope.toRoutingKeys());
```

Comment:
I introduced a check to return before mark it as `waiting` if it was already committed. So it was simply to avoid unnecessary work. I do not believe a comment would be beneficial in this case.

Reply:
Though, on reflection, it is slightly less clear now I've updated the name of the state to `Known` and `ExecutionOrder` (the mapping to `Commit`/`Committed` is less obvious to a reader). I wonder if `CommittedExecutionOrder` would be better here.

Reply:
I just missed the obvious reason it moved which is that if it's committed we aren't waiting for it to commit and the progress log doesn't need to ensure progress (some other progress log on the home shard will do that, guaranteed to know about it because it is committed).

Which could also be the comment. If you have fairly deep knowledge of the whole thing yes it is "obvious"